OpenAMP rsc table sample

arnopo opened the pull request:
This pull request proposes an implementation of the RPMSg protocol based on a resource table. 
The resource table can be use  to declare resources shared between the main and coprocessor. It is integrated in the  co-processor firmware in a specific section. This section can be recognized and parsed by the Linux OS kernel to initialize shared resources such as trace and RPMsg protocol.

Based on this resource table, the objective is to propose a sample that is platform agnostic and that answers to the Linux rpmsg client sample integrated in the Linux kernel distribution.

For time being  this pull request only implements the co-processor side, and could be extended in future to also implement the master part.

arnopo committed:
stm32mp1: update resource table management

Rebase the resource table management to
the new implementation in open-amp module

Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@st.com>

dbkinder commented on samples/subsys/ipc/openamp_rsc_table/README.rst:
```diff
@@ -0,0 +1,42 @@
+********
+
+This application demonstrates how to use OpenAMP with Zephyr based on a resource
+table. It is designed to respond to the Linux rpmsg client sample.
```
Add a reference to the sample:
```suggestion
table. It is designed to respond to the `Linux rpmsg client sample <
https://elixir.bootlin.com/linux/latest/source/samples/rpmsg/rpmsg_client_sample.c>`_.
```
(Assuming I found the right reference.)

dbkinder commented on samples/subsys/ipc/openamp_rsc_table/README.rst:
```diff
@@ -0,0 +1,42 @@
+
+This application demonstrates how to use OpenAMP with Zephyr based on a resource
+table. It is designed to respond to the Linux rpmsg client sample.
+This integration is expected platform independent.
```
Maybe:
```
This sample implementation is platform independent.
```

dbkinder commented on samples/subsys/ipc/openamp_rsc_table/README.rst:
```diff
@@ -0,0 +1,42 @@
+- Parity: None
+- Stop bits: 1
+
+Reset the board and the following message will appear on the corresponding
```
Before you reset the board, don't you need to run the referenced Linux rpmsg client sample somewhere?

arnop2 commented:
@galak Please, could you take a look to this pull request and associated OpenAMP pull request: https://github.com/zephyrproject-rtos/open-amp/pull/2. 
I'm not sure that adding the resource table directly in open-amp module is the right solution, so if you have  better suggestion,  don't hesitate!

dbkinder commented on samples/subsys/ipc/openamp_rsc_table/README.rst:
```diff
@@ -0,0 +1,68 @@
+
+This application demonstrates how to use OpenAMP with Zephyr based on a resource
+table. It is designed to respond to the `Linux rpmsg client sample <https://elixir.bootlin.com/linux/latest/source/samples/rpmsg/rpmsg_client_sample.c>`_.
+This sample implementation is compatible with platforms that embeds a Linux kernel OS on main processor and a Zephyr on the co-processor.
```
Let's change this to reduce the line length and fix misspellings:
```
This sample implementation is compatible with platforms that embed
a Linux kernel OS on the main processor and a Zephyr application on
the co-processor.
```

dbkinder commented on samples/subsys/ipc/openamp_rsc_table/README.rst:
```diff
@@ -0,0 +1,68 @@
+Linux
+------
+
+Enable SAMPLE_RPMSG_CLIENT configuration to build and install the rpmsg_client_sample.ko module on the target.
```
```suggestion
Enable SAMPLE_RPMSG_CLIENT configuration to build and install
the ``rpmsg_client_sample.ko`` module on the target.
```

dbkinder commented on samples/subsys/ipc/openamp_rsc_table/README.rst:
```diff
@@ -0,0 +1,68 @@
+Result on Zephyr console
+-------------------------
+
+The following message will appear on the corresponding Zephyr sconsole:
```
```suggestion
The following message will appear on the corresponding Zephyr console:
```

arnop2 replied:
@dkinder remarks applied, thanks

dbkinder reviewed (approved):
doc changes LTGM, thanks!

erwango reviewed (changes requested):
First batch of comments

arnopo committed:
dts: extend stm32mp1 sram declaration to 300 kB by default

64 kB of memory is reserved for the inter-processor
communication. this makes sense only if RPMsg is used.
Allow to use this memory for firmware data by default.

Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@st.com>

erwango commented on scripts/dts/extract/globals.py:
```diff
@@ -25,7 +25,8 @@
     'zephyr,ccm'   : 'DT_CCM',
-    'zephyr,dtcm'  : 'DT_DTCM'
+    'zephyr,dtcm'  : 'DT_DTCM',
+    'zephyr,ipc_shm' : 'DT_IPC_SHM'
```
Expect conflicts with #17660

erwango replied:
#17660 is now merged, you'll have to report this patch in `scripts/dts/gen_defines.py`

arnop2 replied:
done

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -80,6 +80,7 @@ def main():
     write_addr_size(edt, "zephyr,sram", "SRAM")
     write_addr_size(edt, "zephyr,ccm", "CCM")
     write_addr_size(edt, "zephyr,dtcm", "DTCM")
+    write_addr_size(edt, "zephyr,ipc_shm", "IPC_SHM")
```
Seems okay.

ulfalizer commented on samples/subsys/ipc/openamp_rsc_table/remote/stm32mp157c_dk2.overlay:
```diff
@@ -0,0 +1,24 @@
+	};
+
+	mcusram3: memory1@10040000 {
+		device_type = "memory";
```
Please remove this line (see https://github.com/zephyrproject-rtos/zephyr/pull/17860).

ulfalizer commented on samples/subsys/ipc/openamp_rsc_table/remote/stm32mp157c_dk2.overlay:
```diff
@@ -0,0 +1,24 @@
+
+&mcusram {
+	reg = <0x10000000 DT_SIZE_K(256)>;
+};
```
Missing newline at end of file.

arnop2 replied:
@ulfalizer : remarks taken into account, thanks!

arnopo committed:
samples: add openamp sample relying on resource table

This sample is designed to respond to the Linux
rpmsg sample client.
It should be platform independent and based on the
the integration of a resource table in the elf file.

Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@st.com>

ulfalizer reviewed (approved):
Review comments fixed

arnop2 commented on modules/Kconfig.rsc-table:
```diff
@@ -0,0 +1,27 @@
+	  add the resource table in the generated binary. This table is
+	  compatible with linux remote proc framework and OpenAMP library.
+
+config SC_TABLE_SRC_PATH
```
fixed, thanks

arnopo committed:
lib: open-amp: add helper to add resource table in project

The resource table is needed by the Linux kernel OS
for a rpmsg generic support, but is also recognised by OpenAMP.
This table allows to add trace based on the RAM console
and to support rpmsg protocol.

Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@st.com>

jhedberg commented:
@arnopo could you please rebase this and fix the conflicts? Is this otherwise ready to be merged?

arnopo commented:
@jhedberg 
yes i will rebase it,  
for me it is then ready for merge but would be better first that @galak approve it.

arnopo commented:
@galak: kind reminer. This PR still wait your review :)

jhedberg commented:
This seems to have been neglected for awhile. I just triggered a Shippable re-run. @galak could you take a look?

galak commented on lib/open-amp/rsc_table/Kconfig:
```diff
@@ -0,0 +1,20 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+config RSC_TABLE
```
Let's make this `config OPENAMP_RSC_TABLE`

galak commented on lib/open-amp/rsc_table/Kconfig:
```diff
@@ -0,0 +1,20 @@
+	  add the resource table in the generated binary. This table is
+	  compatible with linux remote proc framework and OpenAMP library.
+
+config RSC_TABLE_NUM_RPMSG_BUFF
```
`config OPENAMP_RSC...`

galak commented on lib/open-amp/rsc_table/README:
```diff
@@ -0,0 +1,22 @@
+
```
nit: remove blank line from start of file.

galak commented on lib/open-amp/rsc_table/resource_table.c:
```diff
@@ -0,0 +1,59 @@
+
+extern char ram_console[];
+
+#define __section_t(S)          __attribute__((__section__(#S)))
```
Can we use Z_GENERIC_SECTION() macro.

arnopo replied:
done

galak commented on samples/subsys/ipc/openamp_rsc_table/remote/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+cmake_minimum_required(VERSION 3.13.1)
```
let's remove the 'remote' dir and move everything up one.

arnopo replied:
done

galak commented on samples/subsys/ipc/openamp_rsc_table/remote/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+project(openamp_rsc_table_remote)
+
+zephyr_compile_definitions(
+  -DMETAL_MAX_DEVICE_REGIONS=2
```
What is `-DMETAL_MAX_DEVICE_REGIONS` for?

arnopo replied:
This is used to define the size of the static table that register the memory region. by default value is 1 ( shared memory for rpmsg). Here a second one is used for the resource table. that's why it is redefined

galak commented on samples/subsys/ipc/openamp_rsc_table/remote/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+
+zephyr_compile_definitions(
+  -DMETAL_MAX_DEVICE_REGIONS=2
+  -DVIRTIO_SLAVE_ONLY
```
ideally we get -DVIRTIO_SLAVE_ONLY this from my PR, but fine for now.

arnopo replied:
Please, could you point out your PR, to help me to understand your comment?

galak replied:
It's this old PR https://github.com/zephyrproject-rtos/zephyr/pull/14750 which I plan on updating.

arnopo replied:
i integrated your commit

galak commented on samples/subsys/ipc/openamp_rsc_table/remote/stm32mp157c_dk2.overlay:
```diff
@@ -0,0 +1,23 @@
+		/*
+		 * shared memory reserved for the inter-processor communication
+		 */
+		zephyr,ipc_shm = &mcusram3;
```
add a `zephyr,ipc` here.

galak commented on soc/arm/st_stm32/stm32mp1/dts_fixup.h:
```diff
@@ -195,6 +195,8 @@
 #define DT_GPIO_STM32_GPIOK_CLOCK_BUS           \
 				DT_ST_STM32_GPIO_5000C000_CLOCK_BUS
 
+#define DT_IPM_DEV		DT_ST_STM32_IPCC_MAILBOX_4C001000_LABEL
```
I want to avoid adding DT_IPM_DEV, see what I did in PR #17553 w/the Kconfig and such to get the params from the device-tree.

galak reviewed (changes requested):
In general looks good, a few updates based on changes I made to PR #17553 that we should reflect here.

galak reviewed (commented):
Do you see having more things in  lib/open-amp/. Wonder if we need the rsc_table dir, or if we can flatten things a little?

arnopo commented:
Thanks @galak for the review, i will push a new version ASAP ( likely tomorrow)

arnopo commented:
> Do you see having more things in lib/open-amp/. Wonder if we need the rsc_table dir, or if we can flatten things a little?

No more things it pipe for the moment, ok i will suppress the rsc_dtable dir

arnopo commented:
> > Do you see having more things in lib/open-amp/. Wonder if we need the rsc_table dir, or if we can flatten things a little?
> 
> No more things it pipe for the moment, ok i will suppress the rsc_dtable dir

rsc_table dir has been suppressed

galak commented on lib/open-amp/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+zephyr_include_directories_ifdef(CONFIG_OPENAMP_RSC_TABLE .)
```
I think we can remove this line now.

arnopo replied:
need to keep it to be able to include resource_table.h in sample

galak commented on samples/subsys/ipc/openamp_rsc_table/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+project(openamp_rsc_table_remote)
+
+zephyr_compile_definitions(
+  -DMETAL_MAX_DEVICE_REGIONS=2
```
forgot to ask, can you add a comment here about why we are setting this to 2.

galak reviewed (commented):
In general looks good, some minor updates.  Should we get PR #14750 merged first, after which I think this should be good.

arnopo commented on lib/open-amp/README:
```diff
@@ -0,0 +1,21 @@
+Description:
```
README file is not standard in zephyr( zephyrbots report), so I will move the description to resource_table.c as a file header comment.

arnopo commented:
Shippable issue seems not linked to this PR:
```
INFO    - 435/782 mps2_an385                tests/net/socket/select/net.socket.select          FAILED Failed (qemu 5.352s)
INFO    - /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/sanity-out/mps2_an385/tests/net/socket/select/net.socket.select/handler.log
ERROR   - *** Booting Zephyr OS build zephyr-v2.2.0-812-g60dd11610e6d  ***
Running test suite socket_select
===================================================================
starting test - test_fd_set
PASS - test_fd_set
===================================================================
starting test - test_select

    Assertion failed at WEST_TOPDIR/zephyr/tests/net/socket/select/src/main.c:115: test_select: (tstamp >= 30U && tstamp <= 30 + FUZZ is false)

```

arnopo commented:
> Shippable issue seems not linked to this PR:
> 
> ```
> INFO    - 435/782 mps2_an385                tests/net/socket/select/net.socket.select          FAILED Failed (qemu 5.352s)
> INFO    - /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/sanity-out/mps2_an385/tests/net/socket/select/net.socket.select/handler.log
> ERROR   - *** Booting Zephyr OS build zephyr-v2.2.0-812-g60dd11610e6d  ***
> Running test suite socket_select
> ===================================================================
> starting test - test_fd_set
> PASS - test_fd_set
> ===================================================================
> starting test - test_select
> 
>     Assertion failed at WEST_TOPDIR/zephyr/tests/net/socket/select/src/main.c:115: test_select: (tstamp >= 30U && tstamp <= 30 + FUZZ is false)
> ```

outdated as solve by a rebase

galak commented on samples/subsys/ipc/openamp_rsc_table/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+target_include_directories(app PRIVATE ${LIBMETAL_INCLUDE_DIR} ${OPENAMP_INCLUDE_DIR} ${PLATFORM_DIR})
+
+target_sources(app PRIVATE src/main_remote.c)
+target_include_directories(app PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/..)
```
Is this correct now that we moved things up one level?

arnopo commented:
@galak, 
i tested your updates and rebase to force  the re-run of some stalled checks,
